query-algebrizer/src/cc.rs

```diff
@@ -0,0 +1,795 @@
+                // know it can only return results if treated as a keyword, and we can treat it as
+                // such.
+                if let Some(ValueType::Ref) = value_type {
+                    if let Some(entid) = self.entid_for_ident(schema, kw) {
+                        self.constrain_column_to_entity(alias.1.clone(), DatomsColumn::Value, entid)
+                    } else {
+                        // A resolution failure means we're done here: this attribute must have an
+                        // entity value.
+                        self.mark_known_empty("Value ident didn't resolve.");
+                        return;
+                    }
+                } else {
+                    unimplemented!();
+                };
+            },
+            PatternValuePlace::Constant(ref c) => {
+                // TODO: don't allocate.
+                let typed_value = c.clone().into_typed_value();
+                if !typed_value.is_congruent_with(value_type) {
+                    // If the attribute and its value don't match, the pattern must fail.
+                    self.mark_known_empty("Value constant not congruent with attribute type.");
+                    return;
+                }
+
+                // TODO: if we don't know the type of the attribute because we don't know the
+                // attribute, we can actually work backwards to the set of appropriate attributes
+                // from the type of the value itself! #292.
+                self.constrain_column_to_constant(alias.1.clone(), DatomsColumn::Value, c.clone().into_typed_value());
+            },
+        }
+
+    }
+
+    pub fn apply_pattern<'s, 'p>(&mut self, schema: &'s Schema, pattern: &'p Pattern) {
+        // For now we only support the default source.
+        match pattern.source {
+            Some(SrcVar::DefaultSrc) | None => (),
+            _ => unimplemented!(),
+        };
+
+        if let Some(alias) = self.alias_table(schema, &pattern) {
+            self.apply_pattern_clause_for_alias(schema, pattern, &alias);
+            self.from.push(alias);
+        } else {
+            // We didn't determine a table, likely because there was a mismatch
+            // between an attribute and a value.
+            // We know we cannot return a result, so we short-circuit here.
+            self.mark_known_empty("Table aliaser couldn't determine a table.");
+        }
+    }
+}
+
+#[cfg(test)]
+mod testing {
+    use super::*;
+
+    fn associate_ident(schema: &mut Schema, i: &str, e: Entid) {
+        schema.entid_map.insert(e, i.to_string());
+        schema.ident_map.insert(i.to_string(), e);
+    }
+
+    fn add_attribute(schema: &mut Schema, e: Entid, a: Attribute) {
+        schema.schema_map.insert(e, a);
+    }
+
+    #[test]
+    fn test_unknown_ident() {
+        let mut cc = ConjoiningClauses::new();
+        let schema = Schema::new();
+
+        cc.apply_pattern(&schema, &Pattern {
```

Comment:
Please use some small harness that expresses the inputs, at least, in EDN.  Trying to parse these tests is challenging.